datafusion/src/physical_plan/file_format/parquet.rs

```diff
@@ -236,32 +237,56 @@ impl ExecutionPlan for ParquetExec {
 
         let adapter = SchemaAdapter::new(self.base_config.file_schema.clone());
 
-        let join_handle = task::spawn_blocking(move || {
-            if let Err(e) = read_partition(
-                object_store.as_ref(),
-                adapter,
-                partition_index,
-                &partition,
-                metrics,
-                &projection,
-                &pruning_predicate,
-                batch_size,
-                response_tx.clone(),
-                limit,
-                partition_col_proj,
-            ) {
-                println!(
+        let join_handle = if projection.is_empty() {
+            task::spawn_blocking(move || {
+                if let Err(e) = read_partition_no_file_columns(
+                    object_store.as_ref(),
+                    &partition,
+                    batch_size,
+                    response_tx.clone(),
```

Comment:
Let us pass down `pruning_predicate` and `metrics`, do row group level filtering in `read_partition_no_file_columns` since it's possible to eliminate the whole parquet file together with its partition column value when all its row groups fail the filter. At the same time, `limit` would also be affected.

Reply:
Row group level filtering is not required. Since only partition columns are queried there will be no filters which will operate on the columns contained within the parquet file. The file list will already be pruned due to the listing logic in ListingTable.  Similarly the ParquetFileMetrics only contain metrics about the evaluation of row group predicates evaluation and filtering I could add the metrics back in but they won't be used for anything.
```
struct ParquetFileMetrics {
    /// Number of times the predicate could not be evaluated
    pub predicate_evaluation_errors: metrics::Count,
    /// Number of row groups pruned using
    pub row_groups_pruned: metrics::Count,
}
```

Reply:
For `SELECT year,month,day FROM t where id > 0`, I think the id>0 filter would pass down to parquet scan even if it's not in the "select" list?


Here's the output of `explain verbose ....`, I make changes to print pushed down filters in ParquetExec as well.
```
+-------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type                                             | plan                                                                                                                                                                                                      |
+-------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| initial_logical_plan                                  | Projection: #t.year, #t.month, #t.day                                                                                                                                                                     |
|                                                       |   Filter: #t.id > Int64(0)                                                                                                                                                                                |
|                                                       |     TableScan: t projection=None                                                                                                                                                                          |
| logical_plan after simplify_expressions               | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after common_sub_expression_eliminate    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after eliminate_limit                    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after projection_push_down               | Projection: #t.year, #t.month, #t.day                                                                                                                                                                     |
|                                                       |   Filter: #t.id > Int64(0)                                                                                                                                                                                |
|                                                       |     TableScan: t projection=Some([0, 11, 12, 13])                                                                                                                                                         |
| logical_plan after filter_push_down                   | Projection: #t.year, #t.month, #t.day                                                                                                                                                                     |
|                                                       |   Filter: #t.id > Int64(0)                                                                                                                                                                                |
|                                                       |     TableScan: t projection=Some([0, 11, 12, 13]), filters=[#t.id > Int64(0)]                                                                                                                             |
| logical_plan after limit_push_down                    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after SingleDistinctAggregationToGroupBy | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after ToApproxPerc                       | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after simplify_expressions               | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after common_sub_expression_eliminate    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after eliminate_limit                    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after projection_push_down               | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after filter_push_down                   | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after limit_push_down                    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after SingleDistinctAggregationToGroupBy | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan after ToApproxPerc                       | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| logical_plan                                          | Projection: #t.year, #t.month, #t.day                                                                                                                                                                     |
|                                                       |   Filter: #t.id > Int64(0)                                                                                                                                                                                |
|                                                       |     TableScan: t projection=Some([0, 11, 12, 13]), filters=[#t.id > Int64(0)]                                                                                                                             |
| initial_physical_plan                                 | ProjectionExec: expr=[year@1 as year, month@2 as month, day@3 as day]                                                                                                                                     |
|                                                       |   FilterExec: CAST(id@0 AS Int64) > 0                                                                                                                                                                     |
|                                                       |     ParquetExec: limit=None, partitions=[year=2021/month=09/day=09/file.parquet, year=2021/month=10/day=09/file.parquet, year=2021/month=10/day=28/file.parquet], filters=CAST(id_max@0 AS Int64) > 0     |
|                                                       |                                                                                                                                                                                                           |
| physical_plan after aggregate_statistics              | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| physical_plan after hash_build_probe_order            | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| physical_plan after coalesce_batches                  | ProjectionExec: expr=[year@1 as year, month@2 as month, day@3 as day]                                                                                                                                     |
|                                                       |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                             |
|                                                       |     FilterExec: CAST(id@0 AS Int64) > 0                                                                                                                                                                   |
|                                                       |       ParquetExec: limit=None, partitions=[year=2021/month=09/day=09/file.parquet, year=2021/month=10/day=09/file.parquet, year=2021/month=10/day=28/file.parquet], filters=CAST(id_max@0 AS Int64) > 0   |
|                                                       |                                                                                                                                                                                                           |
| physical_plan after repartition                       | ProjectionExec: expr=[year@1 as year, month@2 as month, day@3 as day]                                                                                                                                     |
|                                                       |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                             |
|                                                       |     FilterExec: CAST(id@0 AS Int64) > 0                                                                                                                                                                   |
|                                                       |       RepartitionExec: partitioning=RoundRobinBatch(12)                                                                                                                                                   |
|                                                       |         ParquetExec: limit=None, partitions=[year=2021/month=09/day=09/file.parquet, year=2021/month=10/day=09/file.parquet, year=2021/month=10/day=28/file.parquet], filters=CAST(id_max@0 AS Int64) > 0 |
|                                                       |                                                                                                                                                                                                           |
| physical_plan after add_merge_exec                    | SAME TEXT AS ABOVE                                                                                                                                                                                        |
| physical_plan                                         | ProjectionExec: expr=[year@1 as year, month@2 as month, day@3 as day]                                                                                                                                     |
|                                                       |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                             |
|                                                       |     FilterExec: CAST(id@0 AS Int64) > 0                                                                                                                                                                   |
|                                                       |       RepartitionExec: partitioning=RoundRobinBatch(12)                                                                                                                                                   |
|                                                       |         ParquetExec: limit=None, partitions=[year=2021/month=09/day=09/file.parquet, year=2021/month=10/day=09/file.parquet, year=2021/month=10/day=28/file.parquet], filters=CAST(id_max@0 AS Int64) > 0 |
|                                                       |                                                                                                                                                                                                           |
+-------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```

Reply:
From your example above note that the TableScan projects 4 columns and not 3. This means that datafusion is reading 4 columns from the table which is more than the number of partition columns.  
```
Projection: #t.year, #t.month, #t.day                                                                                                                                                                     
   Filter: #t.id > Int64(0)                                                                                                                                                                                
     TableScan: t projection=Some([0, 11, 12, 13]), filters=[#t.id > Int64(0)]
```

This is because datafusion knows that it needs to read the id column from the source to evaluate the query even if it doesn't exist in the final results. 0 is the id column while 11,12, and 13 are the virtual partition columns. This will never hit the code path that I've added since I check if the file projection is empty. This is calculated from the snippet below.
```
let projection = match self.base_config.file_column_projection_indices() {
    Some(proj) => proj,
    None => (0..self.base_config.file_schema.fields().len()).collect(),
};
```
This projection for the example provided will be ```[0]``` and so will be executed using the existing code path. The additional code that I've added will only be executed for the example table provided above if the query only refers to the partition columns year, month, or day. If I modify the test in path_partition.rs to use this query ``` SELECT distinct year,month,day FROM t where id > 0 ``` and add some print statements inside the if statement that decides which function to use, the output is:
```
Non-empty projection. using read_partition
```

And if I revert the query back to  ``` SELECT distinct year,month,day FROM t ``` the output is:
```
Empty projection. using read_partition_no_file_columns
```

Reply:
Thanks for your patience and detailed explanation. I was wrong on this.